app/scripts/lib/multichain-api/wallet-requestPermissions.js

```diff
@@ -0,0 +1,163 @@
+import { isPlainObject } from '@metamask/controller-utils';
+import { invalidParams, MethodNames } from '@metamask/permission-controller';
+import {
+  CaveatTypes,
+  RestrictedMethods,
+} from '../../../../shared/constants/permissions';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from './caip25permissions';
+
+export const requestPermissionsHandler = {
+  methodNames: [MethodNames.requestPermissions],
+  implementation: requestPermissionsImplementation,
+  hookNames: {
+    requestPermissionsForOrigin: true,
+    getPermissionsForOrigin: true,
+    getNetworkConfigurationByNetworkClientId: true,
+    updateCaveat: true,
+    grantPermissions: true,
+  },
+};
+
+/**
+ * Request Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.requestPermissionsForOrigin - The specific method hook needed for this method implementation
+ * @param options.getPermissionsForOrigin
+ * @param options.getNetworkConfigurationByNetworkClientId
+ * @param options.updateCaveat
+ * @param options.grantPermissions
+ * @returns A promise that resolves to nothing
+ */
+async function requestPermissionsImplementation(
+  req,
+  res,
+  _next,
+  end,
+  {
+    requestPermissionsForOrigin,
+    getPermissionsForOrigin,
+    getNetworkConfigurationByNetworkClientId,
+    updateCaveat,
+    grantPermissions,
+  },
+) {
+  const { origin, params } = req;
+
+  if (!Array.isArray(params) || !isPlainObject(params[0])) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const [requestedPermissions] = params;
+  delete requestedPermissions[Caip25EndowmentPermissionName];
+
+  const [_grantedPermissions] = await requestPermissionsForOrigin(
+    requestedPermissions,
+  );
+
+  // caveat values are frozen and must be cloned before modified
+  const grantedPermissions = { ..._grantedPermissions };
+
+  const ethAccountsPermission =
+    grantedPermissions[RestrictedMethods.eth_accounts];
+
+  if (process.env.BARAD_DUR && ethAccountsPermission) {
```

Comment:
isn't this entire API behind this feature flag (`BARAD_DUR`)? So isn't the check redundant?

Reply:
oh nvm this is the adapter loll